groups/etcd_group_service.go

```diff
@@ -0,0 +1,232 @@
+				if err != nil {
+					logger.Log.Warn("[groups] sd: error rebootstrapping lease, will retry in 5 seconds")
+					time.Sleep(5 * time.Second)
+					continue
+				} else {
+					return
+				}
+			}
+		}
+	}
+}
+
+func memberGroupKey(groupName, uid string) string {
+	return "uids/" + uid + "/groups/" + groupName
+}
+
+func groupPrefix(groupName string) string {
+	return "groups/" + groupName
+}
+
+func memberKey(groupName, uid string) string {
+	return groupPrefix(groupName) + "/uids/" + uid
+}
+
+// MemberGroups returns all groups which member takes part
+func (c *EtcdGroupService) MemberGroups(ctx context.Context, uid string) ([]string, error) {
+	prefix := memberGroupKey("", uid)
+	etcdRes, err := clientInstance.Get(ctx, prefix, clientv3.WithKeysOnly(), clientv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+
+	groups := make([]string, etcdRes.Count)
+	for i, kv := range etcdRes.Kvs {
+		groups[i] = string(kv.Key)[len(prefix):]
+	}
+	return groups, nil
+}
+
+// Members returns all member's UID in current group
+func (c *EtcdGroupService) Members(ctx context.Context, groupName string) (map[string]*Payload, error) {
+	prefix := memberKey(groupName, "")
+	etcdRes, err := clientInstance.Get(ctx, prefix, clientv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+
+	members := make(map[string]*Payload, etcdRes.Count)
+	for _, kv := range etcdRes.Kvs {
+		payload := &Payload{}
+		if err = json.Unmarshal(kv.Value, payload); err != nil {
+			return nil, err
+		}
+		members[string(kv.Key)[len(prefix):]] = payload
+	}
+	return members, nil
+}
+
+// Member returns the Payload from User
+func (c *EtcdGroupService) Member(ctx context.Context, groupName, uid string) (*Payload, error) {
+	etcdRes, err := clientInstance.Get(ctx, memberKey(groupName, uid))
+	if err != nil {
+		return nil, err
+	}
+	payload := &Payload{}
+	if err = json.Unmarshal(etcdRes.Kvs[0].Value, payload); err != nil {
+		return nil, err
+	}
+	return payload, nil
+}
+
+// Contains check whether a UID is contained in current group or not
+func (c *EtcdGroupService) Contains(ctx context.Context, groupName, uid string) (bool, error) {
+	etcdRes, err := clientInstance.Get(ctx, memberKey(groupName, uid), clientv3.WithCountOnly())
+	if err != nil {
+		return false, err
+	}
+	return etcdRes.Count > 0, nil
+}
+
+// Add adds UID and payload to group. If the group doesn't exist, it is created
+func (c *EtcdGroupService) Add(ctx context.Context, groupName, uid string, payload *Payload) error {
+	jsonPayload, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+	_, err = clientInstance.Put(ctx, memberKey(groupName, uid), string(jsonPayload), clientv3.WithLease(leaseID))
+	if err != nil {
+		return err
+	}
+	_, err = clientInstance.Put(ctx, memberGroupKey(groupName, uid), "", clientv3.WithLease(leaseID))
+	return err
+}
+
+// Leave removes specified UID from group
+func (c *EtcdGroupService) Leave(ctx context.Context, groupName, uid string) error {
```

Comment:
Leave and Leave all are not the best names. I'd rather use Delete(All) or Remove(All) since you're deleting/removing the members from the group instead of an user leaving a group.

Reply:
Okie Dokie!